### PR TITLE
cudaPackages.cuda-samples: misc bumps and fixups

### PR DESCRIPTION
**Pull request by ConnorBaker**
## Description of changes

- Don't include `cuda-samples` in `cudaPackages_11_7` because there isn't a compatible release.
  - Better to have it absent than to throw and break evaluation in a way that can't easily be overridden.
- Fixes a bug where `backendStdenv` was always taken from the default `cudaPackages`, regardless of which version of `cudaPackages` was trying to build `cuda-samples`.
- Refactors to use `finalAttrs` instead of the `rec` keyword.
- Fixes CMake-related build failure for CUDA Packages 12.2+.
  - NOTE: This does not fix 12.0+ builds; see below.
- Adds CUDA team as maintainers.
- Switches to `autoAddOpenGLRunpathHook`.

> [!WARNING]
> Builds with CUDA Packages 12+ are currently failing with the following:
>
> ```console
> cuda-samples> /nix/store/x1254667lvjd1d760lhz0irxr9q40338-cudatoolkit-12.1.1/bin/nvcc -ccbin g++ -m64 -gencode arch=compute_50,code=compute_50 -o memMapIPCDrv helper_multiprocess.o memMapIpc.o -L/nix/store/x1254667lvjd1d760lhz0irxr9q40338-cudatoolkit-12.1.1/lib64/stubs -lcuda
> cuda-samples> /nix/store/x1254667lvjd1d760lhz0irxr9q40338-cudatoolkit-12.1.1/include/cuda/std/barrier(144): error: function "operator new" cannot be called with the given argument list
> cuda-samples>             argument types are: (unsigned long, cuda::std::__4::__barrier_base<cuda::std::__4::__empty_completion, 2> *)
> cuda-samples>           { new (&__b->__barrier) __barrier_base(__expected); }
> cuda-samples>             ^
> cuda-samples> 1 error detected in the compilation of "tf32TensorCoreGemm.cu".
> cuda-samples> make[1]: *** [Makefile:367: tf32TensorCoreGemm.o] Error 255
> ```
>
> I imagine this is unrelated to this PR -- the only reason it appears now is because this PR allows compilation with newer versions of CUDA. Prior to this PR, the version used was fixed to the version of `cudaPackages` defined at the top level in Nixpkgs, regardless of which `cudaPackages_*_*` was used.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ConnorBaker** [comment]
Result of `nixpkgs-review pr 266115 --extra-nixpkgs-config '{
      allowUnfree = true;
      cudaSupport = true;
      cudaCapabilities = [ "8.9" ];
      packageOverrides = pkgs: { cudaPackages = pkgs.lib.recurseIntoAttrs pkgs.cudaPackages_11_8; };
    }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cudaPackages.cuda-samples</li>
  </ul>
</details>

**Comment by SomeoneSerge** [comment]
> Better to have it absent than to throw and break evaluation in a way that can't easily be overridden.

Is it?

I think you may be right, but I'll try to argue to the opposite.

Hydra handles `broken = true` and gracefully (?) skips the broken derivation. It's kind of similar for `nixpkgs-review`. I did some ugly hacks to mimic that behaviour in https://github.com/SomeoneSerge/nixpkgs-cuda-ci. The convenient thing about it is that `broken = true` is relatively explicit and unambiguous: some human had a look at the code and decided to mark it as "don't bother trying to build this". I say "relatively", because attributes get also classified as "broken" when it's just their transitive dependencies that are explicitly marked.

Interpreting an absent attribute in the CI to me seems more tricky. Say you want to assert that a valid package set should at least attributes from a certain list, and that they should be buildable. E.g. it should have `python3Packages.torchWithCuda`, `python3Packages.jax`, and, well, `cudaPackages.cuda-samples`. Dropping the `cuda-samples` attribute  depending on the release means we need to implement extra logic in the CI that would judge whether the missing attribute is an error or not based on the cuda/nixpkgs version.

The argument maybe doesn't hold for `cuda-samples` specifically because I think it is missing for some releases anyway, but in general I'd say having a more static set of attribute paths is a good thing?

**Comment by SomeoneSerge** [comment]
> `finalAttrs`

😍 

> `buildInputs = [ cudatoolkit ]`

[Is it time](https://youtu.be/JFfkTqvy9uA), since you're at it?

**Comment by ConnorBaker** [comment]
> > Better to have it absent than to throw and break evaluation in a way that can't easily be overridden.
> 
> Is it?

To clarify, it's not "break" in the sense that the derivation has `broken = true` -- I'm replacing `builtins.throw` which would "break" evaluation entirely.

**Comment by ConnorBaker** [comment]
> > `buildInputs = [ cudatoolkit ]`
> 
> [Is it time](https://youtu.be/JFfkTqvy9uA), since you're at it? (Or actually maybe you shouldn't waste the breath, there are more important matters and this otherwise can be merged fast; idk)

I'd prefer to do that in a different PR if that's okay with you :)

**Comment by SomeoneSerge** [comment]
> To clarify, it's not "break" in the sense that the derivation has broken = true -- I'm replacing builtins.throw which would "break" evaluation entirely.

Right, we definitely do not want to `throw`. But `broken = true` is still an alternative to `optionalAttrs`, what do you think about those?

File: pkgs/test/cuda/cuda-samples/generic.nix [hunk 1]
```diff
@@ -1,31 +1,46 @@
...
+    rev = "v${finalAttrs.version}";
     inherit sha256;
   };
 
-  nativeBuildInputs = [ pkg-config addOpenGLRunpath glfw3 freeimage ];
+  nativeBuildInputs = [
+    pkg-config
+    autoAddOpenGLRunpathHook
+    glfw3
+    freeimage
```
Comment (samuela) [inline pkgs/test/cuda/cuda-samples/generic.nix:29]: nit: let's alphabetize these while we're here

Comment (ConnorBaker) [reply pkgs/test/cuda/cuda-samples/generic.nix:29]: Resolved in 83d5298d37fd77d812787950f0e16a16b5ecf1bb

**Comment by ConnorBaker** [comment]
> Right, we definitely do not want to `throw`. But `broken = true` is still an alternative to `optionalAttrs`, what do you think about those?

I think this deserves its own followup, so I made an issue to track it: https://github.com/NixOS/nixpkgs/issues/266475

It's a specific case of how we choose to handle unavailable/unknown comparabilities.

**Pull request merged**